sootty/utils.py

```diff
@@ -10,5 +12,165 @@ def dec2anybase(input, base, width):
         else:
             res += chr(rem - 10 + ord("A"))
         input = int(input / base)
-
+ 
     return res[::-1].zfill(ceil(log(2**width - 1, base)))
+
+def vcdid_hash(s):
+    val = 0
+    s_len = len(s)
+    for i in reversed(range(s_len)):
+        val *= 94
+        val += s[i] - 32
+
+    return val
+
+def vcdid_unhash(value):
+    buf = str()
+    while value:
+        if vmod := value % 94:
+            buf += chr(vmod + 32)
+        else:
+            buf += '~'
+            value -= 94
+        value = int(value / 94)
+    return buf.encode()  # bytes(buf, 'utf-8') or 'ascii'
+
+def evcd_strcpy(src, direction):
+    evcd = b"DUNZduLHXTlh01?FAaBbCcf"
+    vcdi = b"01xz01zzzzzz01xz0011xxz"  # input direction = False
+    vcdo = b"zzzzzz01xz0101xz1x0x01z"  # output direction = True
+    vcd = vcdo if direction else vcdi
+    value = bytes()
+    for vc_bit in src:
+        for i in range(23):
+            if evcd[i] == vc_bit:
+                value += vcd[i:i+1]
+                i -= 1
+                break
+        if i == 22:
+            raise SoottyError("EVCD value error: value change is invalid")
+    return value
+
+def evcd2vcd(stream):
+    # while (line := stream.readline()):
+    #     if (line.startswith(b'$scope') or line.startswith(b'$upscope') or line.startswith(b'$endd') or line.startswith(b'$timescale') or line.startswith(b'$date') or line.startswith(b'$version')):
+    #         vcd_bytes.write(line)
+    #     elif (line.startswith(b'$var')):
+    #         toks = re.split(b' |\t|\n|\r\n', line)
+    buf = stream.read()
+    toks = buf.split()  # compared with re.split(b'\s+', buf), this automatically strip
+    tokit = iter(toks)
+    vcd = BytesIO()
+    vcd_ids = dict()
+    try:
+        while True:
+            tok = next(tokit)
+            # Basic formatter and syntax checker
+            if tok == b'$comment' or tok == b'$date' or tok == b'$timescale' or tok == b'$version' or tok == b'$vcdclose':
+                vcd.write(tok + b' ')
+                while (body := next(tokit)) != b'$end':
```

Comment:
I'd prefer to avoid using the `:=` operator because this feature is only supported by Python >3.8. It could prevent anyone who runs an older version from running the program.